openretina/utils/plotting.py

```diff
@@ -388,3 +400,370 @@ def legend_without_duplicate_labels(ax=None) -> None:
+        saturations = {}
+        for i in range(len(type_boundaries) - 1):
+            start, end = type_boundaries[i], type_boundaries[i + 1]
+            types_in_range = np.where((cell_types >= start) & (cell_types <= end))[0]
+            base_saturation = 0.6 + (i * (0.3 / len(type_boundaries)))  # Vary saturation slightly for each broad type
+            for type_idx in types_in_range:
+                saturations[cell_types[type_idx]] = base_saturation
+
+        # Assign hues based on cell type numbers
+        for cell_type in np.unique(cell_types):
+            # Map cell type to hue (0-1)
+            hue = (cell_type - min_type) / (max_type - min_type)
+            saturation = saturations[cell_type]
+            color_map[cell_type] = (hue, saturation)
+
+    # Pre-compute centroids for each ROI if we're visualizing IDs
+    centroids = {}
+    if visualize_ids:
+        for i in range(n_neurons):
+            if roi_ids is not None and (i + 1) not in roi_ids:
+                continue
+            neuron_mask = roi_mask == -(i + 1)
+            if np.any(neuron_mask):
+                centroids[i] = center_of_mass(neuron_mask)
+
+    def render_frame(frame_idx):
+        """Render a single frame using matplotlib"""
+        # Create figure and axis
+        fig, ax = plt.subplots(figsize=figsize)
+        plt.close()
+
+        # Create frame image (now in RGB)
+        frame_image = np.zeros((*roi_mask.shape, 3))
+
+        # Update each neuron's color based on type and brightness based on activity
+        if roi_ids is not None:
+            for i, roi_id in enumerate(roi_ids):
+                neuron_mask = roi_mask == -roi_id
+                activity_val = activity_normalized[i, frame_idx]
+
+                if cell_types is not None:
+                    # Get pre-computed hue and saturation for this cell type
+                    hue, saturation = color_map[cell_types[i]]
+                    # Convert HSV to RGB
+                    rgb = mcolors.hsv_to_rgb([hue, saturation, activity_val])
+                    frame_image[neuron_mask] = rgb
+                else:
+                    # If no cell types provided, use grayscale
+                    frame_image[neuron_mask] = [activity_val] * 3
+        else:
+            for i in range(n_neurons):
+                neuron_mask = roi_mask == -(i + 1)
+                activity_val = activity_normalized[i, frame_idx]
+
+                if cell_types is not None:
+                    hue, saturation = color_map[cell_types[i]]
+                    rgb = mcolors.hsv_to_rgb([hue, saturation, activity_val])
+                    frame_image[neuron_mask] = rgb
+                else:
+                    frame_image[neuron_mask] = [activity_val] * 3
+
+        # Display frame
+        ax.imshow(frame_image)
+        ax.axis("off")
+
+        # Add neuron ID labels if requested
+        if visualize_ids:
+            for i in range(n_neurons):
+                if i in centroids:
```

Comment:
```suggestion
            for i in [for x in range(n_neurons) if x in centroids]:
               ...
```
minor, but less indentation levels is normally good.